test/blackbox/README.md

```diff
@@ -0,0 +1,28 @@
+# Blackbox tests
+
+These black-box tests provide the majority of the package coverage, to make sure that it works as expected when fully
+wired up and serving HTTP traffic. This is because most of what it does is parsing global state, setting headers,
+rendering output, etc which would be hard and brittle to test with unit or integration tests.
+
+The tests are run against the **built version of the package**. To do this, we use `git export` to build a distribution
+package, then use composer to install that with its dependencies inside a docker container running apache.
+
+Note that this means when working locally, changes to the application code **will not be picked up by the test_subject 
+automatically**. To update the test_subject you will need to **commit your changes** (so that they are included in the
+`git export HEAD`) and then **exit and re-provision the blackbox testing stack**.
+
```

Comment:
Does using `git export` here prevent us using `.gitattributes` to exclude the `/test` and `/.github` dirs from the package dist, so we have to carry all this extra weight around everywhere?

Alternatively - I guess we could exclude these dirs in `.dockerignore` in the projects using this package to achieve a similar effect if needs be

I would also add something to this comment about _why_ we're using git export - is it purely because we have to bring it back in through composer or because we're explicitly trying to test the built package.

Reply:
First off there's a typo/brainfart in this readme that it's `git archive` not `git export`. And I'll clarify that the point of using it is because that then produces the exact same tar that will be produced by GitHub and consumed by composer - e.g. respecting the .gitattrributes and ignored files - so is the key step that allows us to blackbox test the distribution version of the package rather than the files we have checked out locally.

Reply:
Ahh ok, it's all starting to make a lot more sense now.

Reply:
We should probably add a .gitattributes at this point too so we can see what is excluded

Reply:
There's one already, from the initial version https://github.com/ingenerator/microframework/blob/1.x/.gitattributes - `/test` is excluded from the export.